src/consensus/merkle.h

```diff
@@ -90,4 +93,1361 @@ uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated = NULL);
  */
 std::vector<uint256> BlockMerkleBranch(const CBlock& block, uint32_t position);
 
+/*
+ * Each link of a Merkle tree can have one of three values in a proof
+ * object:
+ *
+ *   DESCEND: This link connects to another sub-tree, which must be
+ *     processed. The root of this sub-tree is the hash value of the
+ *     link.
+ *
+ *   VERIFY: This hash value of this link must be provided at
+ *     validation time. Computation of the Merkle root and comparison
+ *     with a reference value provides a batch confirmation as to
+ *     whether ALL the provided VERIFY hashes are correct.
+ *
+ *   SKIP: The hash value of this link is provided as part of the
+ *     proof.
+ */
+enum class MerkleLink : unsigned char { DESCEND, VERIFY, SKIP };
+
+/*
+ * An internal node can have up to eight different structures, the
+ * product of the 3 possible MerkleLink states the left and right
+ * branches can have, with the exception of the {SKIP, SKIP} state
+ * which would be pruned as a SKIP hash in the parent node.
+ *
+ * This means nodes can be represented as a 3-bit integer, and packed
+ * 8 nodes to each 3 byte sequence. The MerkleNode class uses an
+ * unsigned char to represent the unpacked code, whereas the
+ * MerkleNodeReference class is used to access a 3-bit code value
+ * within a packed representation.
+ */
+struct MerkleNode
+{
+    typedef unsigned char code_type;
+
+protected:
+    code_type m_code;
+
+    static const std::array<MerkleLink, 8> m_left_from_code;
+    static const std::array<MerkleLink, 8> m_right_from_code;
+
+    static code_type _get_code(MerkleLink left, MerkleLink right);
+
+public:
+    explicit MerkleNode(MerkleLink left, MerkleLink right) : m_code(_get_code(left, right)) { }
+
+    explicit MerkleNode(code_type code) : m_code(code) { }
+
+    /* Note that a code value of 0 is a {VERIFY, SKIP} node. */
+    MerkleNode() : m_code(0) { }
+
+    /* The default behavior is adequate. */
+    MerkleNode(const MerkleNode&) = default;
+    MerkleNode(MerkleNode&&) = default;
+    MerkleNode& operator=(const MerkleNode&) = default;
+    MerkleNode& operator=(MerkleNode&&) = default;
+
+    /*
+     * Ideally this would perhaps be operator int() and operator=(),
+     * however C++ does not let us mark an assingment operator as
+     * explicit. This unfortunately defeats many of the protections
+     * against bugs that strong typing would give us as any integer or
+     * Boolean value could be mistakenly assigned and interpreted as a
+     * code, and therefore assignable to a MerkleNode, probably
+     * generating a memory access exception if the value is not
+     * between 0 and 7.
+     */
+    inline code_type GetCode() const
+      { return m_code; }
+
+    inline MerkleNode& SetCode(code_type code)
+    {
+        m_code = code;
+        return *this;
```

Comment:
I assume you use the fact this returns itself to nest stuff, but you would normally expect a setter to return nothing, or the value being set, not the object being set, I think. No big deal, but worth pointing out.

Reply:
Yeah it's a consistent design pattern that setters return a reference to themselves. For classes which have more than one setter this allows chaining them in a single line. You can also do something like `some_func(..., MerkleNode().SetCode(...), ...)`.

Doesn't really serve a purpose right now that I'm aware of, but it'd good practice in case the class becomes adds more functionality in the future that might make this idiom useful.